FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,165 @@
+    _size = UserDefaultsBacked(key: "\(defaultsPrefix).model-size")
+    _path = UserDefaultsBacked(key: "\(defaultsPrefix).model-path")
+  }
+}
+
+/// Model info retriever for a model from local user defaults or server.
+class ModelInfoRetriever: NSObject {
+  /// Current Firebase app.
+  var app: FirebaseApp
+  /// Model info associated with model.
+  var modelInfo: ModelInfo?
+  /// Model name.
+  var modelName: String
+  /// Firebase installations.
+  var installations: Installations
+  /// User defaults associated with model.
+  var defaults: UserDefaults
+
+  /// Associate model info retriever with current Firebase app and model name.
+  init(app: FirebaseApp, modelName: String, defaults: UserDefaults = .firebaseMLDefaults) {
+    self.app = app
+    self.modelName = modelName
+    self.defaults = defaults
+    installations = Installations.installations(app: app)
+  }
+
+  /// Construct model fetch base URL.
+  var modelInfoFetchBaseURL: URL {
+    var components = URLComponents()
+    components.scheme = "https"
+    components.host = "firebaseml.googleapis.com"
+    components.path = "/Model"
+    return components.url!
+  }
+
+  /// Build custom model object from model info.
+  func buildModel() -> CustomModel? {
+    /// Build custom model only if model info is filled out, and model file is already on device.
+    guard let info = modelInfo, let path = info.path else { return nil }
+    let model = CustomModel(
+      name: info.name,
+      size: info.size,
+      path: path,
+      hash: info.hash
+    )
+    return model
+  }
+}
+
+/// Extension to handle reading and writing model info to user defaults.
+extension ModelInfoRetriever {
+  /// Model info from server.
+  func retrieveModelInfo(request: URLRequest) {
+    // TODO: Get model info from server and save to user defaults
+  }
+
+  /// FIS token for Firebase app.
+  func authTokenForApp(app: FirebaseApp) {
+    // TODO: Get FIS auth token
+  }
+}
+
+/// Named user defaults for FirebaseML.
+extension UserDefaults {
+  static var firebaseMLDefaults: UserDefaults {
+    let suiteName = "com.google.firebase.ml"
+    return UserDefaults(suiteName: suiteName)!
+  }
+
+  /// For testing: returns a new cleared instance of user defaults.
+  static func getTestInstance() -> UserDefaults {
+    let suiteName = "com.google.firebase.ml.test"
+    let defaults = UserDefaults(suiteName: suiteName)!
+    defaults.removePersistentDomain(forName: suiteName)
+    return defaults
+  }
+}
+
+/// Property initializer for user defaults. Value is always read from or written to a named user defaults store.
+@propertyWrapper struct UserDefaultsBacked<Value> {
+  let key: String
+  let defaultValue: Value
+  let storage: UserDefaults = .firebaseMLDefaults
+
+  var wrappedValue: Value {
+    get {
+      let value = storage.value(forKey: key) as? Value
+      return value ?? defaultValue
+    }
+    set {
+      guard let optional = newValue as Optional?, optional != nil else {
```

Comment:
What I'm checking here is if `newValue` is nil - is this the right way to do that?

Reply:
I wonder if something like the following will work:
```
if let newValue = newValue else { 
  storage.setValue(newValue, forKey: key)
} else {
  storage.removeObject(forKey: key)
}
```

Reply:
I don't think this will work. I was playing around in a playground and remember seeing something like this before and came up with a different (somewhat messy?) solution:

```
private protocol OptionalProtocol: ExpressibleByNilLiteral {
  var isNil: Bool { get }
}

extension Optional: OptionalProtocol {
  public var isNil: Bool {
    switch self {
    case .some(_): return false
    case .none: return true
    }
  }
}
```

This is just saying that there's a protocol that `nil` "conforms" to and Optional does as well, then the check can be against the protocol:

```
if let opt = newValue as? OptionalProtocol, opt.isNil {
  // Value is nil, remove it.
} else {
  // Value is not nil.
}
```

Does that make sense? It's not the nicest but it works as far as I can tell :P

Reply:
Makes sense, I updated the code to look more like this.